[V2V] Use authentication_check instead of verify credentials

djberg96 opened the pull request:
This alters the `ConversionHost#eligible?` method so that it calls the `authentication_check` method instead of `verify_credentials` directly.

Internally the `authentication_check` method calls the `verify_credentials` method, and then updates the authentications object, including the `last_valid_on` field. We use that field to skip the ssh connection check if it was valid within the last 15 minutes. This will help reduce unnecessary ssh connections if the method happens to be called multiple times in a short period of time, e.g. `InfraConversionThrottler.start_conversions`. Plus it reduces the chance of unexpected timeout errors, etc.

Update: The 15 minute idea was dropped in favor of just using the `authentication_status_ok?` method that already exists in the Authentication mixin.

~~WIP until I can get some tests added.~~

Some other updates included in this PR:

* I set the default auth_type to "v2v" in the `verify_credentials` method.

* I changed the resources in the throttler specs to be provider-specific.

* I added an "RSpec" to the outer describe for future rspec 4 compliance.

* I added the "v2v" tag to the throttler specs.

* I changed "Max Transformation Runners" to "MaxTransformationRunners" because it was causing a deprecation warning:

`DEPRECATION WARNING: Invalid custom attribute: 'Max Transformation Runners'.  A custom attribute name must begin with a letter (a-z, but also letters with diacritical marks and non-Latin letters) or an underscore (_). Subsequent characters can be letters, underscores, digits (0-9), or dollar signs ($) (called from block (3 levels) in <top (required)> at /Users/dberger/Dev/manageiq-djberg96/spec/lib/infra_conversion_throttler_spec.rb:31)`

Not exactly related, but does obliquely help with https://github.com/ManageIQ/manageiq/pull/18880 a bit.

Addresses https://bugzilla.redhat.com/show_bug.cgi?id=1723420

djberg96 commented:
@fdupont-redhat What do you think?

djberg96 commented:
@miq-bot add_labels transformation, changelog/yes, hammer/yes

djberg96 commented:
@miq-bot remove_label wip

djberg96 commented:
@miq-bot add_reviewers @agrare @kbrock

djberg96 commented:
@miq-bot add_reviewers @agrare, @kbrock

fdupont-redhat commented:
@miq-bot add-reviewer @agree
@miq-bot add-reviewer @kbrock

kbrock commented on lib/infra_conversion_throttler.rb:
```diff
@@ -3,7 +3,7 @@ def self.start_conversions
       running = ems.conversion_hosts.inject(0) { |sum, ch| sum + ch.active_tasks.count }
       $log&.debug("There are currently #{running} conversion hosts running.")
-      slots = (ems.miq_custom_get('Max Transformation Runners') || Settings.transformation.limits.max_concurrent_tasks_per_ems).to_i - running
+      slots = (ems.miq_custom_get('MaxTransformationRunners') || Settings.transformation.limits.max_concurrent_tasks_per_ems).to_i - running
```
thnx - this was causing warnings for me

do we have existing code or customers that uses this old variable name?

djberg96 replied:
Good question. @fdupont-redhat will this affect existing automate code for customers?

fdupont-redhat replied:
Nope. It's used in the throttler only. But the documentation has to be updated to reflect that change.

djberg96 replied:
@fdupont-redhat Thanks for the update.

kbrock commented on spec/lib/infra_conversion_throttler_spec.rb:
```diff
@@ -25,18 +25,20 @@
       allow(conversion_host1).to receive(:check_ssh_connection).and_return(true)
       allow(conversion_host2).to receive(:check_ssh_connection).and_return(true)
+      allow(conversion_host1).to receive(:authentication_check).and_return([true, 'passed'])
+      allow(conversion_host2).to receive(:authentication_check).and_return([true, 'passed'])
```
keep this

but this stubbing is skipping actually testing the code.
try and keep stubbing to a method that provides just an external interface (e.g.: connect to ssh - or maybe even connect to the database, though that is suspect)

djberg96 replied:
Those particular tests are just checking the throttling, not sure what you mean.

kbrock commented on app/models/conversion_host.rb:
```diff
@@ -91,7 +95,7 @@ def verify_credentials(auth_type = nil, options = {})
   #
   def eligible?
-    source_transport_method.present? && verify_credentials && check_concurrent_tasks
+    source_transport_method.present? && authentication_check('v2v').first && check_concurrent_tasks
```
this throws me a little.

In `verify_credentials`, the connect code uses v2v authentications || the first one. This logic looks different.

I liked having some high level function in here like `verify_credentials`. This middle clause has me wondering if `authentication_check.first` is right. Or what about the check when there are no authentications that seems to happen. So those low level question seems out of place in this high level method.

djberg96 replied:
Internally `authentication_check` calls `verify_credentials`, it's essentially an interface method that classes should define on their own.

kbrock replied:
I read the method, and understand why this is true.
But reading the old code for some reason felt higher level

could we add a readability method?

```ruby
def authentication_valid_for?(*args)
  authentication_check(*args).first
end
```

djberg96 replied:
I get what you're saying but I'd like to keep it minimalist for now, since that would have to be added to the Authentication mixin.

djberg96 replied:
I think @agrare's suggestion to use `authentication_status_ok?` accomplishes what you were after, and I've updated the PR to use it.

fdupont-redhat commented:
@miq-bot add-reviewer @agrare

fdupont-redhat commented:
@djberg96 shouldn't verify_credentials return false if authentication was not invalid within the last X minutes ? X could be 2, to not attempt connection too frequently, but also not leave the host as ineligible during too long.

djberg96 commented:
@fdupont-redhat Not sure I'm following. Do you want me to add logic that checks `Authentication#last_invalid_on` as well?

djberg96 commented on app/models/conversion_host.rb:
```diff
@@ -68,6 +68,10 @@ def verify_credentials(auth_type = nil, options = {})
 
+      # Don't connect again if the authentication was valid within the last 15
+      # minutes. This helps reduce unnecessary ssh connections.
+      return true if auth.last_valid_on && auth.last_valid_on > 15.minutes.ago
```
Eh, I thought we should validate the auth type before checking if it's valid.

You could argue that this would be a nice method to have baked into either the Authentication model or the mixin, but that's a separate PR.

It only uses `check_ssh_connection` if there's no associated authentication.

djberg96 replied:
I think the use of `authentication_status_ok?` essentially resolves your concerns here.

djberg96 commented on app/models/conversion_host.rb:
```diff
@@ -46,7 +46,7 @@ class ConversionHost < ApplicationRecord
   #
-  def verify_credentials(auth_type = nil, options = {})
+  def verify_credentials(auth_type = 'v2v', options = {})
     if authentications.empty?
```
Nah, we still want it to make a best guess via `authentication_type` below.

djberg96 replied:
And, we'll want the auth_type to be 'v2v' for the `authentication_status_ok?` method or it could fail.

fdupont-redhat commented:
> @fdupont-redhat Not sure I'm following. Do you want me to add logic that checks `Authentication#last_invalid_on` as well?

Yes. Otherwise for a host with invalid authentication, we'll do SSH connections until it becomes valid. I think we should evince it for a period of time. Maybe we could have the same check interval for both, no ? What about 5 minutes ? That would still lower the number of SSH connections, while limiting the risk of assigning an invalid host to a task.

djberg96 commented:
@fdupont-redhat I split the difference and changed it to 10 minutes. Is that ok? 5 minutes seems too short.

I also updated it so that if it was invalid in the last 10 minutes it will still retry, even if it was also valid within that time. I updated the specs, too.

agrare commented on app/models/conversion_host.rb:
```diff
@@ -91,7 +97,7 @@ def verify_credentials(auth_type = nil, options = {})
   #
   def eligible?
-    source_transport_method.present? && verify_credentials && check_concurrent_tasks
+    source_transport_method.present? && authentication_check('v2v').first && check_concurrent_tasks
```
@djberg96 could you use `authentication_status_ok?` instead?  Then you wouldn't need the logic to not check the credentials if they were checked recently.  It would just rely on the scheduled auth check.

djberg96 replied:
@agrare Good call, I've updated the PR. I think this addresses an issue that @kbrock had as well.

djberg96 commented:
@kbrock Are you ok with it now?

djberg96 commented:
@miq-bot remove_label hammer/yes

djberg96 commented:
@miq-bot add_label hammer/no

agrare commented on app/models/conversion_host.rb:
```diff
@@ -91,7 +94,7 @@ def verify_credentials(auth_type = nil, options = {})
   #
   def eligible?
-    source_transport_method.present? && verify_credentials && check_concurrent_tasks
+    source_transport_method.present? && authentication_check('v2v').first && check_concurrent_tasks
```
Hm I was assuming that you would call `authentication_status_ok?` here...is anyone calling `verify_credentials` anymore now ?  I don't see it being used in this file.

djberg96 replied:
It gets called internally via the `authentication_check` (via a private method within that method).

https://github.com/ManageIQ/manageiq/blob/master/app/models/mixins/authentication_mixin.rb#L349

agrare replied:
Ah okay :+1:

agrare commented:
This looks better, @kbrock you good with this?
